socketshark/receiver.py

```diff
@@ -79,29 +83,35 @@ async def ping_handler(self):
             if wait:
                 wait.cancel()
             if not self._stop:
-                self.shark.log.exception('unhandled exception in ping handler')
+                self.shark.log.debug('redis ping handler cancelled')
         except Exception:
             self.shark.log.exception('unhandled exception in ping handler')
+            self._stop = True
         finally:
-            await self.stop()
+            if self._stop:
+                await self.stop()
 
     async def reader(self, once=False):
         self._stop = False
         try:
             ping_handler = asyncio.ensure_future(self.ping_handler())
-            result = await self._reader(once=once)
+            tasks = [
+                self._reader(connection, once=once)
```

Comment:
Is it possible that – if `once` is false, we're gonna be stuck waiting for a message from the 1st Redis (which nobody publishes to anymore) while the messages from the other Redises don't get processed?

Reply:
Each `_reader` task is [gathered](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) into a single awaitable. The coroutines are given equal priority so one cannot block the other.